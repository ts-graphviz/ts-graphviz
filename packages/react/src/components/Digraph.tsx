import React, { FC, useMemo } from 'react';
import PropTypes from 'prop-types';
import * as gv from 'ts-graphviz';
import { RootClusterContext } from '../contexts/RootClusterContext';
import { ClusterContext } from '../contexts/ClusterContext';
import { useGraphvizContext } from '../hooks/useContext';

type Props = {
  id?: string;
};

export const Digraph: FC<Props> = ({ children, id }) => {
  const context = useGraphvizContext();
  const g = useMemo(() => new gv.Digraph(context, id), [context, id]);
  return (
    <RootClusterContext.Provider value={g}>
      <ClusterContext.Provider value={g}>{children}</ClusterContext.Provider>
    </RootClusterContext.Provider>
  );
};

Digraph.displayName = 'Digraph';

Digraph.defaultProps = {
  id: undefined,
};

Digraph.propTypes = {
  id: PropTypes.string,
};
