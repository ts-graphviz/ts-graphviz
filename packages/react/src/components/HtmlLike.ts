/* eslint-disable no-undef */
import { AttributesValue } from 'ts-graphviz';

export type TableProps = {
  ALIGN?: 'CENTER' | 'LEFT' | 'RIGHT'; // "CENTER|LEFT|RIGHT"
  BGCOLOR?: AttributesValue; // "color"
  BORDER?: AttributesValue; // "value"
  CELLBORDER?: AttributesValue; // "value"
  CELLPADDING?: AttributesValue; // "value"
  CELLSPACING?: AttributesValue; // "value"
  COLOR?: AttributesValue; // "color"
  COLUMNS?: AttributesValue; // "value"
  FIXEDSIZE?: true; // "FALSE|TRUE"
  GRADIENTANGLE?: AttributesValue; // "value"
  HEIGHT?: AttributesValue; // "value"
  HREF?: AttributesValue; // "value"
  ID?: AttributesValue; // "value"
  PORT?: AttributesValue; // "portName"
  ROWS?: AttributesValue; // "value"
  SIDES?: AttributesValue; // "value"
  STYLE?: AttributesValue; // "value"
  TARGET?: AttributesValue; // "value"
  TITLE?: AttributesValue; // "value"
  TOOLTIP?: AttributesValue; // "value"
  VALIGN?: 'MIDDLE' | 'BOTTOM' | 'TOP'; // "MIDDLE|BOTTOM|TOP"
  WIDTH?: AttributesValue; // "value"
};

type NoAttributes = {};

export type TrProps = NoAttributes;

export type TdProps = {
  ALIGN?: 'CENTER' | 'LEFT' | 'RIGHT' | 'TEXT'; // "CENTER|LEFT|RIGHT|TEXT"
  BALIGN?: 'CENTER' | 'LEFT' | 'RIGHT'; // "CENTER|LEFT|RIGHT"
  BGCOLOR?: AttributesValue; // "color"
  BORDER?: AttributesValue; // "value"
  CELLPADDING?: AttributesValue; // "value"
  CELLSPACING?: AttributesValue; // "value"
  COLOR?: AttributesValue; // "color"
  COLSPAN?: AttributesValue; // "value"
  FIXEDSIZE?: boolean; // "FALSE|TRUE"
  GRADIENTANGLE?: AttributesValue; // "value"
  HEIGHT?: AttributesValue; // "value"
  HREF?: AttributesValue; // "value"
  ID?: AttributesValue; // "value"
  PORT?: AttributesValue; // "portName"
  ROWSPAN?: AttributesValue; // "value"
  SIDES?: AttributesValue; // "value"
  STYLE?: AttributesValue; // "value"
  TARGET?: AttributesValue; // "value"
  TITLE?: AttributesValue; // "value"
  TOOLTIP?: AttributesValue; // "value"
  VALIGN?: 'MIDDLE' | 'BOTTOM' | 'TOP'; // "MIDDLE|BOTTOM|TOP"
  WIDTH?: AttributesValue; // "value"
};

export type FontProps = {
  color?: AttributesValue; // "color"
  face?: AttributesValue; // "fontname"
  point?: AttributesValue; // SIZE="value"
};

export type BrProps = {
  ALIGN?: 'CENTER' | 'LEFT' | 'RIGHT'; // "CENTER|LEFT|RIGHT"
};

export type ImgProps = {
  SCALE?: boolean | 'WIDTH' | 'HEIGHT' | 'BOTH'; // "FALSE|TRUE|WIDTH|HEIGHT|BOTH"
  SRC?: AttributesValue; // "value"
};

export type IProps = NoAttributes;

export type BProps = NoAttributes;

export type UProps = NoAttributes;

export type OProps = NoAttributes;
export type SubProps = NoAttributes;

export type SupProps = NoAttributes;
export type SProps = NoAttributes;
export type HrProps = NoAttributes;
export type VrProps = NoAttributes;

export enum DOT {
  PORT = 'dot-port',
  TABLE = 'dot-table',
  TR = 'dot-tr',
  TD = 'dot-td',
  FONT = 'dot-font',
  BR = 'dot-br',
  IMG = 'dot-img',
  I = 'dot-i',
  B = 'dot-b',
  U = 'dot-u',
  O = 'dot-o',
  SUB = 'dot-sub',
  SUP = 'dot-sup',
  S = 'dot-s',
  HR = 'dot-hr',
  VR = 'dot-vr',
}

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    interface IntrinsicElements {
      [DOT.PORT]: { children: string };
      [DOT.TABLE]: React.PropsWithChildren<TableProps>;
      [DOT.TR]: React.PropsWithChildren<TrProps>;
      [DOT.TD]: React.PropsWithChildren<TdProps>;
      [DOT.FONT]: React.PropsWithChildren<FontProps>;
      [DOT.BR]: BrProps;
      [DOT.IMG]: ImgProps;
      [DOT.I]: React.PropsWithChildren<IProps>;
      [DOT.B]: React.PropsWithChildren<BProps>;
      [DOT.U]: React.PropsWithChildren<UProps>;
      [DOT.O]: React.PropsWithChildren<OProps>;
      [DOT.SUB]: React.PropsWithChildren<SubProps>;
      [DOT.SUP]: React.PropsWithChildren<SupProps>;
      [DOT.S]: React.PropsWithChildren<SProps>;
      [DOT.HR]: HrProps;
      [DOT.VR]: VrProps;
    }
  }
}
