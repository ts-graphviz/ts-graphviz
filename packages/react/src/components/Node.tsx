import React, { FC, ReactElement } from 'react';
import PropTypes from 'prop-types';
import { useNode, NodeProps } from '../hooks/use-node';
import { useRenderedID } from '../hooks/use-rendered-id';

type Props = Omit<NodeProps, 'label' | 'xlabel'> & {
  label?: ReactElement | string;
  xlabel?: ReactElement | string;
};

export const Node: FC<Props> = ({ children, label, xlabel, ...props }) => {
  const renderedLabel = useRenderedID(label);
  const renderedXlabel = useRenderedID(xlabel);

  if (renderedLabel !== undefined) Object.assign(props, { label: renderedLabel });
  if (renderedXlabel !== undefined) Object.assign(props, { xlabel: renderedXlabel });
  useNode(props);
  return <>{children}</>;
};

Node.displayName = 'Node';

Node.propTypes = {
  id: PropTypes.string.isRequired,
  comment: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  xlabel: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
};

Node.defaultProps = {
  comment: undefined,
  label: undefined,
  xlabel: undefined,
};
