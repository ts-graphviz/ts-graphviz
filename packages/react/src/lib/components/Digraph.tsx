import { type FC, useEffect } from 'react';
import { CurrentGraph } from '../contexts/CurrentGraph.js';
import { GraphContainer } from '../contexts/GraphContainer.js';
import { useDigraph } from '../hooks/useDigraph.js';
import { useGraphContainer } from '../hooks/useGraphContainer.js';
import { useGraphMap } from '../hooks/useGraphMap.js';
import { useRenderedID } from '../hooks/useRenderedID.js';
import type { RootGraphProps } from '../types.js';

/**
 * `Digraph` component.
 */
export const Digraph: FC<RootGraphProps> = ({
  children,
  label,
  ...options
}) => {
  const container = useGraphContainer();
  if (container !== null) {
    throw Error(
      'RootCluster is duplicated.\nUse only one of Digraph and Graph.',
    );
  }
  const renderedLabel = useRenderedID(label);
  if (renderedLabel !== undefined)
    Object.assign(options, { label: renderedLabel });
  const digraph = useDigraph(options);
  const clusters = useGraphMap();
  useEffect(() => {
    if (digraph.id !== undefined) {
      clusters.set(digraph.id, digraph);
    }
  }, [clusters, digraph]);
  return (
    <GraphContainer.Provider value={digraph}>
      <CurrentGraph.Provider value={digraph}>{children}</CurrentGraph.Provider>
    </GraphContainer.Provider>
  );
};

Digraph.displayName = 'Digraph';
