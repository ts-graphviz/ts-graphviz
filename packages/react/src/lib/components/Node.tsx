import type { FC } from 'react';
import { useNode } from '../hooks/useNode.js';
import { useRenderedID } from '../hooks/useRenderedID.js';
import type { NodeProps } from '../types.js';

/**
 * `Node` component.
 */
export const Node: FC<NodeProps> = ({ id, label, xlabel, ...options }) => {
  const renderedLabel = useRenderedID(label);
  const renderedXlabel = useRenderedID(xlabel);

  if (renderedLabel !== undefined)
    Object.assign(options, { label: renderedLabel });
  if (renderedXlabel !== undefined)
    Object.assign(options, { xlabel: renderedXlabel });
  useNode(id, options);
  return null;
};

Node.displayName = 'Node';
