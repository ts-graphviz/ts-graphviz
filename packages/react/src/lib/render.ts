import { type ReactElement, createElement } from 'react';
import { type GraphBaseModel, type RootGraphModel, toDot } from 'ts-graphviz';

import { CurrentGraph } from './contexts/CurrentGraph.js';
import { GraphContainer } from './contexts/GraphContainer.js';
import { GraphMap } from './contexts/GraphMap.js';
import { type Context, GraphvizContext } from './contexts/GraphvizContext.js';
import { reconciler } from './reconciler.js';

const noop = (): void => undefined;

function clusterMap(
  graph?: GraphBaseModel,
  map: Map<string, GraphBaseModel> = new Map(),
): Map<string, GraphBaseModel> {
  if (graph) {
    if (graph.id) {
      map.set(graph.id, graph);
    }
    for (const s of graph.subgraphs) {
      clusterMap(s, map);
    }
  }
  return map;
}

/**
 * Convert the given element to Graphviz model.
 *
 * @example Example of giving a cluster as a container with the second argument.
 *
 * ```tsx
 * import React, { FC } from 'react';
 * import { digraph, toDot } from 'ts-graphviz';
 * import { Node, Subgraph, render, Edge } from '@ts-graphviz/react';
 *
 * const Example: FC = () => (
 *   <>
 *     <Node id="a" />
 *
 *     <Subgraph id="my_cluster">
 *       <Node id="b" />
 *     </Subgraph>
 *     <Edge targets={['b', 'a']} />
 *   </>
 * );
 *
 * const G = digraph((g) => render(<Example />, g));
 * console.log(toDot(G));
 * // digraph {
 * //   "a";
 * //   subgraph "my_cluster" {
 * //     "b";
 * //   }
 * //   "b" -> "a";
 * // }
 * ```
 */
export function render<T extends GraphBaseModel>(
  element: ReactElement,
  container?: T,
): T {
  const context: Context<T> = { container };
  reconciler.updateContainer(
    createElement(
      GraphvizContext.Provider,
      { value: context },
      createElement(
        GraphMap.Provider,
        { value: clusterMap(container) },
        createElement(
          GraphContainer.Provider,
          { value: container ?? null },
          container
            ? createElement(
                CurrentGraph.Provider,
                { value: container },
                element,
              )
            : element,
        ),
      ),
    ),
    reconciler.createContainer(
      {},
      0,
      null,
      true,
      null,
      '@ts-graphviz/react',
      noop,
      null,
    ),
    null,
    noop,
  );
  if (!context.container) {
    throw Error(
      'There are no clusters of container(Subgraph, Digraph, Graph).',
    );
  }
  return context.container;
}

/**
 * Converts the given element to DOT language and returns it.
 *
 * @example
 *
 * ```tsx
 * import React, { FC } from 'react';
 * import { Digraph, Node, Subgraph, renderToDot, Edge } from '@ts-graphviz/react';
 *
 * const Example: FC = () => (
 *   <Digraph>
 *     <Node id="a" />
 *     <Subgraph id="my_cluster">
 *       <Node id="b" />
 *     </Subgraph>
 *     <Edge targets={['b', 'a']} />
 *   </Digraph>
 * );
 *
 * const dot = renderToDot(<Example />);
 * console.log(dot);
 * // digraph {
 * //   "a";
 * //   subgraph "my_cluster" {
 * //     "b";
 * //   }
 * //   "b" -> "a";
 * // }
 * ```
 *
 * @returns Rendered dot string
 */
export function renderToDot(
  element: ReactElement,
  container?: RootGraphModel,
): string {
  return toDot(render(element, container));
}
