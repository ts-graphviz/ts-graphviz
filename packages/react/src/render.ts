// import { ReactElement } from 'react';
// import instance from './instance';

// export function render<Props>(tree: ReactElement<Props>) {
//   return instance().render(tree);
// }
