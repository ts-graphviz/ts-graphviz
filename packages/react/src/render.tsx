import { createElement, type ReactElement, startTransition } from 'react';
import { type GraphBaseModel, toDot } from 'ts-graphviz';

import { CurrentGraph } from './contexts/CurrentGraph.js';
import { GraphContainer } from './contexts/GraphContainer.js';
import { GraphMap } from './contexts/GraphMap.js';
import { type Context, GraphvizContext } from './contexts/GraphvizContext.js';
import { reconciler } from './reconciler.js';

/**
 * Rendering result containing the graph and any metadata
 */
export interface RenderResult<T extends GraphBaseModel> {
  graph: T;
  cleanup?: () => void;
}

/**
 * Rendering options for controlling the rendering behavior
 */
export interface RenderOptions<T extends GraphBaseModel = GraphBaseModel> {
  /**
   * Container graph to render into. When provided, the element will be rendered
   * as children of this container.
   */
  container?: T;
  /**
   * Enable concurrent rendering with startTransition for improved performance
   * @default true
   */
  concurrent?: boolean;
  /**
   * Timeout for async rendering in milliseconds
   * @default 5000
   */
  timeout?: number;
  /**
   * Called when an error is not caught by any error boundary
   * React 19 error handling pattern
   */
  onUncaughtError?: (
    error: Error,
    errorInfo: { componentStack?: string },
  ) => void;
  /**
   * Called when an error is caught by an error boundary
   * React 19 error handling pattern
   */
  onCaughtError?: (
    error: Error,
    errorInfo: { componentStack?: string },
  ) => void;
}

const noop = (): void => undefined;

function clusterMap(
  graph?: GraphBaseModel,
  map: Map<string, GraphBaseModel> = new Map(),
): Map<string, GraphBaseModel> {
  if (graph) {
    if (graph.id) {
      map.set(graph.id, graph);
    }
    for (const s of graph.subgraphs) {
      clusterMap(s, map);
    }
  }
  return map;
}

/**
 * Creates the wrapped React element with all necessary context providers
 */
function createWrappedElement<T extends GraphBaseModel>(
  element: ReactElement,
  context: Context<T>,
  container?: T,
): ReactElement {
  return (
    <GraphvizContext value={context}>
      <GraphMap value={clusterMap(container)}>
        <GraphContainer value={container ?? null}>
          {container ? (
            <CurrentGraph value={container}>{element}</CurrentGraph>
          ) : (
            element
          )}
        </GraphContainer>
      </GraphMap>
    </GraphvizContext>
  );
}

/**
 * Creates a fiber root for rendering with React 19 error handling
 */
function createFiberRoot(containerInfo: any = {}, options: RenderOptions = {}) {
  const { onUncaughtError = noop } = options;

  return reconciler.createContainer(
    containerInfo,
    1, // tag - ConcurrentRoot = 1 for concurrent mode (React 19 default)
    null, // hydrationCallbacks
    false, // isStrictMode
    null, // concurrentUpdatesByDefaultOverride
    '@ts-graphviz/react', // identifierPrefix
    onUncaughtError ? (error: Error) => onUncaughtError(error, {}) : () => {}, // onRecoverableError - simplified for React reconciler compatibility
    null, // transitionCallbacks
  );
}

/**
 * Convert the given element to Graphviz model.
 *
 * @example Example of giving a cluster as a container in options.
 *
 * ```tsx
 * import React, { FC } from 'react';
 * import { digraph, toDot } from 'ts-graphviz';
 * import { Node, Subgraph, render, Edge } from '@ts-graphviz/react';
 *
 * const Example: FC = () => (
 *   <>
 *     <Node id="a" />
 *
 *     <Subgraph id="my_cluster">
 *       <Node id="b" />
 *     </Subgraph>
 *     <Edge targets={['b', 'a']} />
 *   </>
 * );
 *
 * const G = digraph(async (g) => {
 *   const result = await render(<Example />, { container: g });
 *   return result.graph;
 * });
 * console.log(toDot(G));
 * // digraph {
 * //   "a";
 * //   subgraph "my_cluster" {
 * //     "b";
 * //   }
 * //   "b" -> "a";
 * // }
 * ```
 */

/**
 * Internal rendering function without concurrent wrapper
 */
async function renderInternal<T extends GraphBaseModel>(
  element: ReactElement,
  options: RenderOptions<T> = {},
): Promise<RenderResult<T>> {
  const { container, timeout = 5000 } = options;
  const context: Context<T> = { container };
  const containerInfo: any = { context };
  const fiberRoot = createFiberRoot(containerInfo, options);
  const wrappedElement = createWrappedElement(element, context, container);

  return new Promise<RenderResult<T>>((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      reject(new Error(`Rendering timeout after ${timeout}ms`));
    }, timeout);

    try {
      reconciler.updateContainer(wrappedElement, fiberRoot, null, () => {
        clearTimeout(timeoutId);

        // Check both places for the container
        const finalContainer =
          containerInfo.context?.container || context.container;

        if (finalContainer) {
          resolve({
            graph: finalContainer,
            cleanup: () => {
              // Cleanup resources if needed
              reconciler.updateContainer(null, fiberRoot, null, noop);
            },
          });
        } else {
          reject(
            new Error(
              'There are no clusters of container(Subgraph, Digraph, Graph).',
            ),
          );
        }
      });
    } catch (error) {
      clearTimeout(timeoutId);
      reject(error);
    }
  });
}

/**
 * Renders a React element asynchronously.
 * This is the primary rendering method with optional concurrent features.
 */
export async function render<T extends GraphBaseModel>(
  element: ReactElement,
  options: RenderOptions<T> = {},
): Promise<RenderResult<T>> {
  const { concurrent = true } = options;

  if (concurrent) {
    return new Promise<RenderResult<T>>((resolve, reject) => {
      startTransition(() => {
        renderInternal(element, options).then(resolve).catch(reject);
      });
    });
  }
  return renderInternal(element, options);
}

/**
 * Converts the given element to DOT language asynchronously.
 * This is the primary DOT rendering method.
 *
 * @example
 * ```tsx
 * import React, { FC } from 'react';
 * import { Digraph, Node, Subgraph, renderToDot, Edge } from '@ts-graphviz/react';
 *
 * const Example: FC = () => (
 *   <Digraph>
 *     <Node id="a" />
 *     <Subgraph id="my_cluster">
 *       <Node id="b" />
 *     </Subgraph>
 *     <Edge targets={['b', 'a']} />
 *   </Digraph>
 * );
 *
 * const dot = await renderToDot(<Example />);
 * console.log(dot);
 * ```
 */
export async function renderToDot(
  element: ReactElement,
  options: RenderOptions = {},
): Promise<string> {
  const result = await render(element, options);
  try {
    return toDot(result.graph as any);
  } finally {
    // Cleanup resources
    result.cleanup?.();
  }
}
