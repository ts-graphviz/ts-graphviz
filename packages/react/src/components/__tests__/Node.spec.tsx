import React from 'react';
import 'jest-graphviz';
import { Digraph } from '../Digraph';
import { Node } from '../Node';
import { DOT } from '../HtmlLike';
import { renderToDot } from '../../renderer/render';

describe('Node', () => {
  test('pass without optional props and render correctly', () => {
    const dot = renderToDot(
      <Digraph>
        <Node id="aaa" />
      </Digraph>,
    );
    expect(dot).toBeValidDotAndMatchSnapshot();
  });

  test('pass label with string and render correctly', () => {
    const dot = renderToDot(
      <Digraph>
        <Node id="aaa" label="label test" />
      </Digraph>,
    );
    expect(dot).toBeValidDotAndMatchSnapshot();
  });

  test('pass label with HTMLLike ReactElement and render correctly', () => {
    const dot = renderToDot(
      <Digraph>
        <Node
          id="aaa"
          label={
            <DOT.TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0">
              <DOT.TR>
                <DOT.TD>left</DOT.TD>
                <DOT.TD PORT="m">middle</DOT.TD>
                <DOT.TD PORT="r">right</DOT.TD>
              </DOT.TR>
            </DOT.TABLE>
          }
        />
      </Digraph>,
    );
    expect(dot).toBeValidDotAndMatchSnapshot();
  });

  test('pass xlabel with string and render correctly', () => {
    const dot = renderToDot(
      <Digraph>
        <Node id="aaa" xlabel="xlabel test" />
      </Digraph>,
    );
    expect(dot).toBeValidDotAndMatchSnapshot();
  });

  test('pass xlabel with HTMLLike ReactElement and render correctly', () => {
    const dot = renderToDot(
      <Digraph>
        <Node
          id="aaa"
          xlabel={
            <DOT.TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0">
              <DOT.TR>
                <DOT.TD>left</DOT.TD>
                <DOT.TD PORT="m">middle</DOT.TD>
                <DOT.TD PORT="r">right</DOT.TD>
              </DOT.TR>
            </DOT.TABLE>
          }
        />
      </Digraph>,
    );
    expect(dot).toBeValidDotAndMatchSnapshot();
  });
});
