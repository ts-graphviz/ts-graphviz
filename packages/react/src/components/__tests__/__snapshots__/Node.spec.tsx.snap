// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node pass label with HTMLLike ReactElement and render correctly 1`] = `
digraph {
  "aaa" [
    label = <<TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0"><TR><TD>left</TD><TD PORT="m">middle</TD><TD PORT="r">right</TD></TR></TABLE>>,
  ];
}
`;

exports[`Node pass label with string and render correctly 1`] = `
digraph {
  "aaa" [
    label = "label test",
  ];
}
`;

exports[`Node pass without optional props and render correctly 1`] = `
digraph {
  "aaa";
}
`;

exports[`Node pass xlabel with HTMLLike ReactElement and render correctly 1`] = `
digraph {
  "aaa" [
    xlabel = <<TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0"><TR><TD>left</TD><TD PORT="m">middle</TD><TD PORT="r">right</TD></TR></TABLE>>,
  ];
}
`;

exports[`Node pass xlabel with string and render correctly 1`] = `
digraph {
  "aaa" [
    xlabel = "xlabel test",
  ];
}
`;
